### PR TITLE
Implementation of dl-4-tsc Keras classifiers

### PR DESCRIPTION
**Pull request by James-Large**
#### References
* [Project page](https://github.com/alan-turing-institute/sktime/projects/5)
* [Start of potential wiki entry](https://github.com/James-Large/sprint-dlsetup)

#### What does this implement/fix? Explain your changes.
* Initial implementation/interfacing with the Keras networks in [dl-4-tsc](https://github.com/hfawaz/dl-4-tsc) under [sktime.contrib.deeplearning_based](https://github.com/James-Large/sktime/tree/dl4tsc/sktime/contrib/deeplearning_based)
* Default parameters and architectures of all networks maintained for the initial intention of reproducing results from the accompanying [paper](https://link.springer.com/article/10.1007%2Fs10618-019-00619-1) to the above 
* Exact implementation code has also been largely kept the same, except for changes required to fit into the desired API (original was purely for academic experimental code, e.g. preprocessing steps applied to test set alongside the train. For sktime, this has been seperated into predict_proba()) 
* The CNN has had attempts to be generalised, however, in order to allow for tuning via scikit's grid- and randomized- SearchCV
* Basic unit tests added under [sktime/classifiers/tests/test_dl4tscnetworks.py](https://github.com/James-Large/sktime/blob/dl4tsc/sktime/classifiers/tests/test_dl4tscnetworks.py)

In addition, the column ensembler implemented by @ABostrom is included within this PR

* @ABostrom to explain further if needed, but allows for ensembling hetero- or homogeneously over the dimensions of a multivariate problem
* Implementation under [sktime/contrib/column_ensembler.p](https://github.com/James-Large/sktime/blob/dl4tsc/sktime/contrib/column_ensembler.py)
* Testing under [sktime/classifiers/tests/test_columnensembler.py](https://github.com/James-Large/sktime/blob/dl4tsc/sktime/classifiers/tests/test_columnensembler.py)

Minor changes: 

* experiments.py edited to get working locally, can be discarded if desired @James-Large 
* JapaneseVowels multivariate dataset added for datasets base @ABostrom 
* not None check added to label encoder inversing in baseclassifier, however no longer needed, can be discarded if desired @James-Large 

#### Any other comments?

I'm making this PR now to get the code looked at in general and to gather wanted changes - not expecting this to get accepted as is. Time needs to be spent at some point expanding comments and docstrings throughout, and the other networks need to be parameterised similar to the CNN to allow tuning depending on design decisions collectively made on how tuning should be implemented in sktime

Further there does exist some kind of memory leak in these models, specifically those that do some form of internal parameter tuning MCNN, and TLENET (perhaps the others too, however much slower). The exact source and fix I can't figure out right now, that will require extra time. Running a single model on a single dataset seems to be fine. Running many in a single execution eventually leads to an error with very little explanation, but which is almost certainly due to an unexpected/unintended OOM, i.e. a leak. Some links so far into my investigation, but again need more time to fix if possible: 

* Old, but might be a problem with keras itself and how it handles it's tf sessions
    * [1](https://github.com/keras-team/keras/issues/2102), where downgrading Keras may fix: 
        * [2](https://github.com/keras-team/keras/issues/2102#issuecomment-410823700)
        * [3](https://github.com/keras-team/keras/issues/2102#issuecomment-418888790)
* Switching to theano or downgrading is an option... 
* Might also be a problem not with keras specifically, but running keras within pycharm...
    * [4](https://intellij-support.jetbrains.com/hc/en-us/community/posts/360000015484-exit-code-1073741819-0xC0000005-)
    * [5](https://stackoverflow.com/questions/46745523/keras-model-failing-to-compile-exit-code-1073741819-0xc0000005)
    * leading to... [6](https://youtrack.jetbrains.com/issue/PY-17069)
* Running multiple MCNNs in pycharm leads to errors/exit codes, however running from e.g. anaconda prompt looks like it leaks memory much slower, but still eventually fails

Lastly, there is the question of managing the dependencies. Dependencies are: keras, keras-contrib, tensorflow OR tensorflow-gpu and CUDNN(must be installed separately by user). These are not insignificant, and would be the main reason for delaying this PR while we figure this out

### CODE REVIEW

**Comment by mloning** [comment]
Regarding the dependencies, I can think of two options: 
1. Include the deep learning models (estimators as well as transformers as in the shapelet pipeline that @dguijo is working on) in the main package and have keras or tensorflow as a dependency, 
2. Have a separate extension package (`sktime-dl`) with the dependency-heavy models with a dependency on sktime

Any other options? Given the difficulties of installing tensorflow/keras, I think I prefer option two.

**Comment by James-Large** [comment]
Option 2 and separating it into sktime-dl is reasonable I think, even if not permanently. Don't think there's much of a workaround for getting the user to install CUDA/CUDNN for gpu usage separately anyway, so might as well have it as a separate entity with the provision that if you specifically want to use these models, there's a burden on your end to setting them up, but the rest of the package is fine.

**Comment by fkiraly** [comment]
Yes, also strongly supporting option 2, after what I had to go through on my Microsoft machine.

**Comment by TonyBagnall** [comment]
we have the same issue with cython. irrc someone suggested a solution which may apply to keras too in terms of managing dependencies.  Anyone remember what it was called? grendel or somesuch?

**Comment by TonyBagnall** [comment]
and I would imagine that deep learning would play a significant roll in forecasting?

**Comment by James-Large** [comment]
Might be thinking of conda-forge, think @ABostrom suggested it? Not looked into myself yet. Or gradle is the automated build tool, that's a @goastler question

**Comment by mloning** [comment]
* Agreed to create separate repo (@jasonlines) for deep learning methods with keras dependencies.
* Please create separate PR for non-deep-learning stuff like column ensembling.
